[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64241

When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Error before this diff when a comm was aborted:
```
NCCL communicator was aborted on rank 1.
```

With this diff:
```
NCCL communicator was aborted on rank 1. Original reason for failure was: Work: WorkNCCL(OpType=ALLREDUCE, TensorShape=[1], Timeout(ms)=2000) timed out in call to wait().
```

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm

When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -91,12 +95,18 @@ namespace c10d {
 std::string ncclGetErrorWithVersion(ncclResult_t error);
+// Need this wrapper because ProcessGroupNCCL calls needs to call into C10D_NCCL_CHECK
+// with failureReason as an optional argument.
+void c10d_nccl_check_wrapper(ncclResult_t cmd, c10::optional<std::string> failureReason = c10::nullopt);
```
nit: Would it be cleaner to just pass null in all other `C10D_NCCL_CHECK` calls instead of having a wrapper?

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -150,26 +160,38 @@ class NCCLComm {
-      TORCH_CHECK(false,
-          "NCCL communicator was aborted on rank " + std::to_string(rank_) +
-          ".");
+      std::string abortMsg = "NCCL communicator was aborted on rank " + std::to_string(rank_) + ".";
+      if (commFailureReason_ != c10::nullopt) {
+        abortMsg += " . Original reason for failure was: " + *commFailureReason_;
+      }
+      TORCH_CHECK(false, abortMsg);
```
Can we use `fmt::format` here instead?

Separate question - My understanding was that users were seeing errors on the communicator before the work clean up thread actually killed the communicator. But it seems like here we already check for that and used to raise an appropriate exception? Does this mean the failure came probably from other collectives in-flight rather than a new collective call?

rohan-varma replied:
Yeah, although here with this PR we enhance it with the reason for failure if we try to access a bad communicator.

It's sort of hard to root cause exactly where this failure is coming from as it can stem from several places, but internal logging shows that this happens pretty frequently and the `workCleanupThread` raises this exception (either due to a timeout or an actual NCCL comm. issue).

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -150,26 +160,38 @@ class NCCLComm {
   }
 
   ncclComm_t getNcclComm() {
-    std::unique_lock<std::mutex> lock(mutex_);
```
Why did we remove the lock?

rohan-varma replied:
Oops, must've been an accident, not sure how that got removed

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -150,26 +160,38 @@ class NCCLComm {
     }
 
-    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_));
+    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_), commFailureReason_);
```
This is using the member variable `commFailureReason_` instead of the one passed in?

rohan-varma replied:
Oh right, it is not set until L193

rohan-varma replied:
Although, I don't think it really matters. Passing in `commFailureReason_` here is not strictly necessary, as we are doing a `ncclCommAbort` i.e. trying to abort the communicator. Wrapping it in C10d_nccl_check will only cause `commFailureReason_` to be raised if something went wrong in aborting the communicator, which is a case I don't think I've seen.

rohan-varma replied:
Actually, I am not completely sure how we should think about passing in `commFailureReason` to c10d_nccl_check. The main motivation is to ensure all appropriate cases can consider this error when running through the `getNcclErrorDetailStr` error formatter, although a lot of these calls to C10D_NCCL_CHECK are NCCL APIs that don't really have to do with collectives (i.e. ncclCommInitRank, ncclCommAbort, doesn't really have to do anything with the collective itself). 

The calls to C10D_NCCL_CHECK that are indeed collectives (such as the ones in ProcessGroupNCCL::collective) should probably have commFailureReason passed in to ensure we raise the appropriate error if PG work encountered an error though. For the rest of the calls, the benefit is that we'd raise a PyTorch/PG NCCL error where possible instead of a generic NCCL error, which might be better for debugability since we can see what went wrong in our application with regard to timeouts etc.

pritamdamania87 replied:
I think we should pass in the `commFailureReason` to all calls which deal with the communicator irrespective of whether or not it is a collective. Because once the communicator is aborted, other nccl calls which might not necessarily be collectives (ex: ncclCommAbort) might probably fail. As a result, I do think we need to pass in the local `commFailureReason` here and not the member variable.

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -150,26 +160,38 @@ class NCCLComm {
+    // Set true failure reason if provided by ProcessGroupNCCL (e.g. work
+    // timeout)
+    if (commFailureReason != c10::nullopt) {
+      commFailureReason_ = commFailureReason;
```
nit: We can just do `commFailureReason_ = commFailureReason` without the conditional?

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -367,8 +368,10 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
         // if throwing timed out excepiton without aborting nccl communicators
         // here, it was observed that CUDA GPU will have 100% utilization and
         // can not run new events successfully.
+        std::stringstream ss;
+        ss << "Work: " << *this << " timed out in call to wait().";
```
nit: Use fmt:format

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -533,7 +536,9 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
       for (const auto& ncclComm : ncclComms) {
-        ncclComm->ncclCommAbort();
+        std::string abortReason =
+            c10::str("Process Group destroyed on rank ", rank_);
```
nit: fmt::format

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -367,8 +368,10 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
+        ss << "Work: " << *this << " timed out in call to wait().";
         for (const auto& ncclComm : ncclComms_) {
-          ncclComm->ncclCommAbort();
+          ncclComm->ncclCommAbort(ss.str());
```
Don't we need to check for `commFailureReason_` to be set on line 388 and appropriately return that error instead of the time out?

Also, shouldn't we update `checkAndSetException` to first check for `commFailureReason_` first and use that exception instead?

rohan-varma replied:
This part of the code is for NCCL_BLOCKING_WAIT and we are in a section where the work has timed out, so isn't the root cause of aborting the communicator a timeout and we should pass that reason into `ncclCommAbort`? The goal here is to basically make `ncclCommAbort` aware of why we are aborting it and in this case it is a work timeout.

> Also, shouldn't we update checkAndSetException to first check for commFailureReason_ first and use that exception instead?

Yes, looks like we should do this.

rohan-varma replied:
Yep, `ProcessGroupNCCL::WorkNCCL::checkAndSetException()` will try to grab the `commFailureReason` from the NCCL communicator if it exists, via `checkForNcclErrorsInternal`.

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Error before this diff when a comm was aborted:
```
NCCL communicator was aborted on rank 1.
```

With this diff:
```
NCCL communicator was aborted on rank 1. Original reason for failure was: Work: WorkNCCL(OpType=ALLREDUCE, TensorShape=[1], Timeout(ms)=2000) timed out in call to wait().
```

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

cbalioglu commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -60,14 +64,14 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error) {
 
 // Macro to throw on a non-successful NCCL return value.
-#define C10D_NCCL_CHECK(cmd)                                                  \
+#define C10D_NCCL_CHECK(cmd, failureReason)                                                  \
```
Well there are two potential optimizations:

1) The obvious one: declare the function as inline like `[[noreturn]] inline void c10dNcclCheck(...)`
2) I would also avoid using an infinite loop (`while (0)`) and use a real condition (in this case `while (result != ncclSuccess)`). Even modern compilers can have difficulty analyzing `while(true)` loops and might fail to inline the function.

Of course it needs to be measured if this is a performance critical code, but my gut feeling says that there should be no performance difference between a macro and an inline function for this particular case (meaning the function will be inlined in the majority of the use cases)

cbalioglu reviewed (approved):
LGTM

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Error before this diff when a comm was aborted:
```
NCCL communicator was aborted on rank 1.
```

With this diff:
```
NCCL communicator was aborted on rank 1. Original reason for failure was: Work: WorkNCCL(OpType=ALLREDUCE, TensorShape=[1], Timeout(ms)=2000) timed out in call to wait().
```

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]

rohan-varma commented:
It looks like fmt is only enable for libtorch_python sources: https://github.com/pytorch/pytorch/pull/37560/files, but ProcessGroupNCCL/NCCLUtils is not part of libtorch_python as far as I can tell, hence the build errors. 

Filed an issue to refactor and fix this: https://github.com/pytorch/pytorch/issues/64604, will revert to c10::str for now.

rohan-varma committed:
Update on "[c10d] Provide failure reason from ProcessGroup when aborting NCCL comm"


When things go wrong PG NCCL aborts nccl communicators via `ncclCommAbort`, but one issues is that often the error can be set to `ncclSystemError` (see  https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L176) when that might not be the true cause of the issue and the actual issue is that some prior work timed out, communicator was aborted on other rank, etc.

This results in a lot of confusion when debugging jobs with a large no. of processes as the current message for ncclSystemError is not very informative: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L22

The fix here is to pass in a string exception message from PG NCCL down to `NCCLUtils` which will aim to raise that as the actual issue and not the confusing `ncclSystemError` message.

Error before this diff when a comm was aborted:
```
NCCL communicator was aborted on rank 1.
```

With this diff:
```
NCCL communicator was aborted on rank 1. Original reason for failure was: Work: WorkNCCL(OpType=ALLREDUCE, TensorShape=[1], Timeout(ms)=2000) timed out in call to wait().
```

Differential Revision: [D30658855](https://our.internmc.facebook.com/intern/diff/D30658855/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30658855/)!

[ghstack-poisoned]
